packages/style-engine/class-wp-style-engine.php

```diff
@@ -261,29 +365,75 @@ public function generate( $block_styles ) {
+	 * @param string|array $style_value      A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array        $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
 	 *
 	 * @return array The class name for the added style.
 	 */
-	protected static function get_css_rules( $style_value, $style_property ) {
+	protected static function get_css_rules( $style_value, $style_definition ) {
 		$rules = array();
 
 		if ( ! $style_value ) {
 			return $rules;
 		}
 
+		$style_properties = $style_definition['properties'];
+
 		// We assume box model-like properties.
 		if ( is_array( $style_value ) ) {
 			foreach ( $style_value as $key => $value ) {
-				$rules[ "$style_property-$key" ] = $value;
+				$side_property           = strtr( $style_properties['sides'], array( '$side' => _wp_to_kebab_case( $key ) ) );
+				$rules[ $side_property ] = $value;
 			}
 		} else {
-			$rules[ $style_property ] = $style_value;
+			$rules[ $style_properties['default'] ] = $style_value;
+		}
+
+		return $rules;
+	}
+
+	/**
+	 * Style value parser that returns a CSS ruleset for style groups that have 'top', 'right', 'bottom', 'left' keys.
+	 * E.g., `border.top{color|width|style}.
+	 *
+	 * @param array $style_value      A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array $style_definition A single style definition from BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @return array The class name for the added style.
+	 */
+	protected static function get_css_side_rules( $style_value, $style_definition ) {
+		$rules = array();
+
+		if ( ! is_array( $style_value ) || empty( $style_value ) ) {
+			return $rules;
 		}
 
+		foreach ( $style_value as $css_property => $value ) {
+			if ( empty( $value ) ) {
+				continue;
+			}
+			// The first item in the style definition path array tells us the style property, e.g., "border".
+			// We use this to get a corresponding CSS style definition such as "color" or "width" from the same group.
+			$side_style_definition_path = array( $style_definition['path'][0], $css_property );
+			$side_style_definition      = _wp_array_get( self::BLOCK_STYLE_DEFINITIONS_METADATA, $side_style_definition_path, null );
+
+			if ( $side_style_definition && isset( $side_style_definition['properties']['sides'] ) ) {
+				// Set a CSS var if there is a valid preset value.
+				$slug = isset( $style_definition['css_vars'][ $css_property ] ) ? static::get_slug_from_preset_value( $value, $css_property ) : null;
+				if ( $slug ) {
+					$css_var = strtr(
+						$style_definition['css_vars'][ $css_property ],
+						array( '$slug' => $slug )
+					);
+					$value   = "var($css_var)";
+				}
+				// The second item in the style definition path array refers to the side property, e.g., "top".
+				$side_css_property           = strtr( $side_style_definition['properties']['sides'], array( '$side' => $style_definition['path'][1] ) );
```

Comment:
In this case, since we're explicitly in a function for dealing with sides, the `$side` placeholder feels a bit more natural to me, but just a follow-up from my other comment, would it be worth using `%s` and `sprintf()`, or was there a particular reason for using `strtr` for substitution instead? (Apologies if I'm missing context!)

Reply:
No context, just an opinionated approach to suit a particular day, mood, and stellar alignment 😄 

I'll refactor to use `%s` and `sprintf()` 👍